File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        maxLen = 0
        n = len(s)
        l = 0

        hashmap = {}

        for r in range(n):
            if s[r] in hashmap and hashmap[s[r]] >= l:
                l = hashmap[s[r]] + 1

            hashmap[s[r]] = r
            maxLen = max(maxLen, r - l + 1)
            
        return maxLen

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let map = new Map()
    let maxLen = 0
    let left = 0;

    for(let right = 0; right < s.length; right++) {
        let char = s[right]
        if(map.has(char))
            left = Math.max(left, map.get(char) + 1)

        map.set(char, right)
        maxLen = Math.max(maxLen, right - left + 1)
    }

    return maxLen
};